package swift.protocol;
option java_package="com.alibaba.search.swift.protocol";

// encode rule:
// LRUXX
// L indicates level, INFO = 1, WARN = 2, FATAL = 3;
// R indicates role. ALL = 0, ADMIN = 1, BROKER = 2, CLIENT = 3;
// U indicates return to USER or SYS. USER = 1, SYS = 2;
// XX can assign any value as you like.

enum ErrorCode {
/////////////////////////////////
//        COMMON X00XX
/////////////////////////////////
    ERROR_NONE = 0;
    ERROR_UNKNOWN = 1;

    ERROR_INFO_START = 10000;
    ERROR_WARN_START = 20000;
    ERROR_FATAL_START = 30000;
    ERROR_WORKER_DEAD = 30002;
    ERROR_DECOMPRESS_MESSAGE = 3004;
/////////////////////////////////
//        ADMIN ERROR X1XXX
/////////////////////////////////
    // USER ERROR X11XX
    ERROR_ADMIN_TOPIC_HAS_EXISTED = 21101;
    ERROR_ADMIN_TOPIC_NOT_EXISTED = 21102;
    ERROR_ADMIN_INVALID_PARAMETER = 21103;
    ERROR_ADMIN_NOT_LEADER = 21104;
    ERROR_ADMIN_PARTITION_NOT_EXISTED = 21105;
    ERROR_ADMIN_UNKNOWN_COMMAND = 21106;
    ERROR_ADMIN_IP_SET_INVALID = 21107;
    ERROR_ADMIN_INVALID_CONFIG = 21108;
    ERROR_BROKER_IS_UPGRADING = 21109;
    ERROR_UPGRADE_BROKER_BEFORE_ADMIN = 21110;
    ERROR_WRITE_BROKER_TARGET_VERSION = 21111;
    ERROR_UPGRADE_VERSION_SAME_AS_CURRENT = 21112;
    ERROR_ADMIN_APPLICATION_ID_CHANGED = 21113;
    ERROR_CONFIG_IS_ROLLBACKING = 21114;
    ERROR_NOT_NEED_ROLLBACK = 21115;
    ERROR_WRITE_ROLLBACK_FILE = 21116;
    ERROR_UPGRADE_ROLE_VERSION = 21117;
    ERROR_ADMIN_WAIT_IN_QUEUE_TIMEOUT = 21118;
    ERROR_ADMIN_SCHEMA_NOT_FOUND = 21119;
    ERROR_ADMIN_SCHEMA_ALREADY_EXIST = 21120;
    ERROR_ADMIN_TOPIC_NOT_SUPPORT_SCHEMA = 21121;
    ERROR_ADMIN_SCHEMA_VERSION_EXCEED = 21122;
    ERROR_ADMIN_CALC_BROKER_STATUS = 21123;
    ERROR_ADMIN_ROLE_NOT_FOUND = 21124;
    ERROR_ADMIN_SEALED_TOPIC_CANNOT_MODIFY = 21125;
    ERROR_ADMIN_CHG_PART_TASK_NOT_FINISH = 21126;
    ERROR_ADMIN_MASTER_VERSION_ERROR = 21127;
    // ERROR_ADMIN_CALC_MERGER_STATUS = 21128;
    // ERROR_WRITE_MERGER_TARGET_VERSION = 21129;
    // ERROR_MERGER_IS_UPGRADING = 21130;
    ERROR_ADMIN_OPERATION_NOT_PERMITTED = 21131;
    ERROR_ADMIN_AUTHENTICATION_FAILED = 21132;
    ERROR_ADMIN_NOT_MASTER = 21133;

    // SYS ERROR X12XX
    ERROR_ADMIN_NOT_ENOUGH_RESOURCE = 21201;
    ERROR_ADMIN_OPERATION_FAILED = 21202;
    ERROR_ADMIN_CLEAR_OLD_TOPIC_DATA_FAILED = 21203;
    ERROR_ADMIN_SYS_NOT_INIT = 31201;
    ERROR_ADMIN_NOT_IMPLEMENTED = 31202;

/////////////////////////////////
//       BROKER ERROR X2XXX
/////////////////////////////////
    // USER ERROR X21XX
    ERROR_BROKER_FAIL = 12101;
    ERROR_BROKER_STOPPED = 12102;
    ERROR_BROKER_BUSY = 12103;
    ERROR_BROKER_TOPIC_PARTITION_NOT_FOUND = 12104;
    ERROR_BROKER_TOPIC_PARTITION_NOT_RUNNING = 12105;
    ERROR_BROKER_REQUEST_INVALID = 12106;
    // getMessage, getMaxMessageId
    ERROR_BROKER_NO_DATA = 12107;
    ERROR_BROKER_TOPIC_NOT_SUPPORT_FIELD_FILTER = 12108;
    ERROR_BROKER_INIT_FIELD_FILTER_FAILD = 12109;
    ERROR_BROKER_SESSION_CHANGED = 12110;
    ERROR_BROKER_PERMISSION_DENIED = 12111;

    // getMessage
    ERROR_BROKER_SOME_MESSAGE_LOST = 22101;
    // sendMessage
    ERROR_BROKER_MSG_LENGTH_EXCEEDED = 22102;

    // getMinMessageIdByTime
    ERROR_BROKER_TIMESTAMP_TOO_LATEST = 22103;
    ERROR_BROKER_MESSAGE_FORMAT_INVALID = 22104;
    // push work item
    ERROR_BROKER_PUSH_WORK_ITEM = 22105;
    // sealed topic
    ERROR_TOPIC_SEALED = 22106;
    ERROR_SEALED_TOPIC_READ_FINISH = 22107;
    ERROR_BROKER_WRITE_VERSION_INVALID = 22108;

    // SYS ERROR X22XX
    ERROR_BROKER_REPEATED_ADD_TOPIC_PARTITION = 22201;
    ERROR_BROKER_OPEN_FILE = 22202;
    ERROR_BROKER_WRITE_FILE = 22203;
    ERROR_BROKER_COMMIT_FILE = 22204;
    ERROR_BROKER_CLOSE_FILE = 22205;
    ERROR_BROKER_READ_FILE = 22206;
    ERROR_BROKER_FS_OPERATE = 22207;
    ERROR_BROKER_NO_DATA_IN_MEM = 22211;
    ERROR_BROKER_RECOVER_FILE_MESSAGE_NO_MORE_MEMORY = 22212;
    ERROR_BROKER_CREATE_THREAD = 32201;
    ERROR_BROKER_MKDIR = 32202;
    ERROR_BROKER_DATA_DIR_EMPTY = 32203;
    ERROR_BROKER_INLINE_FILE_INVALID = 32204;
    ERROR_BROKER_INLINE_VERSION_INVALID = 32205;
    ERROR_BROKER_INVALID_USER = 32206;

///////////////////////////////////
//       CLIENT ERROR X3XXX
///////////////////////////////////
    // SYS ERROR X32XX
    ERROR_CLIENT_INIT_TWICE = 13201;
    ERROR_CLIENT_INIT_INVALID_PARAMETERS = 13202;
    ERROR_CLIENT_RPC_CONNECT = 13203;
    ERROR_CLIENT_GET_ADMIN_INFO_FAILED = 13204;
    ERROR_CLIENT_GET_BROKER_ADDRESS_FAILED = 13205;
    ERROR_CLIENT_INVALID_RESPONSE = 13206;
    ERROR_CLIENT_SYS_STOPPED = 13207;
    ERROR_CLIENT_ARPC_ERROR = 13208;
    ERROR_CLIENT_READ_MESSAGE_TIMEOUT = 13209;
    ERROR_CLIENT_SEND_BUFFER_FULL = 13210;
    ERROR_CLIENT_SOME_MESSAGE_SEND_FAILED = 13211;
    ERROR_CLIENT_INIT_FAILED = 13212;
    ERROR_CLIENT_NO_MORE_MESSAGE = 13213;
    ERROR_CLIENT_INVALID_PARAMETERS = 13214;
    ERROR_CLIENT_NOT_INITED = 13215;
    ERROR_CLIENT_INVALID_PARTITIONID = 13216;
    ERROR_CLIENT_EOF = 13217;
    ERROR_CLIENT_EXCEED_TIME_STAMP_LIMIT = 13219;
    ERROR_CLIENT_LOAD_SCHEMA_FAIL = 13220;
    ERROR_CLIENT_SCHEMA_MSG_INVALID = 13221;
    ERROR_CLIENT_SCHEMA_VERSION_DIFF = 13222;
    ERROR_CLIENT_SCHEMA_READ_VERSION = 13223;
    ERROR_CLIENT_SCHEMA_MULTIREADER_NOT_SUPPORT = 13224;
    ERROR_CLIENT_LOGIC_TOPIC_EMPTY_PHYSIC = 13225;
    ERROR_CLIENT_PHYSIC_TOPIC_SWITCH_NOT_READY = 13226;
    ERROR_CLIENT_NOT_IMPLEMENTED = 13227;
    ERROR_CLIENT_LOGIC_TOPIC_NOT_IN_BROKER = 13228;
    ERROR_CLIENT_ASYNC_WRITE_TIMEOUT = 13229;
    ERROR_CLIENT_UNSUPPORTED_METHOD = 13230;
    ERROR_CLIENT_ASYNC_WRITE_BUFFER_FULL = 13231;
    ERROR_CLIENT_ARPC_TIMEOUT = 13232;
}

message ErrorInfo {
    optional ErrorCode errCode = 1 [default = ERROR_NONE];
    optional string errMsg = 2;
}

enum ErrorLevel {
    ERROR_LEVEL_INFO = 0;
    ERROR_LEVEL_WARN = 1;
    ERROR_LEVEL_FATAL = 2;
}

message CommonResponse {
    optional ErrorInfo errorInfo = 1;
}
